lens-labels/src/Lens/Labels.hs

```diff
@@ -0,0 +1,106 @@
+{- | A lens library that integrates with OverloadedLabels.
+
+Unlike the `lens` package (and others), lenses are defined as a newtype
+instead of a type synonym, to avoid overlapping with other OverloadedLabel
+instances.  However, the `LensFn` and `runLens` functions allow converting
+between the two types; for example:
+
+> LensFn :: Control.Lens.LensLike f s t a b -> Lens.Labels.LensLike f s t a b
+> runLens :: Lens.Labels.LensLike f s t a b -> Control.Lens.LensLike f s t a b
+
+TODO: support more general optic types (e.g., prisms).
+-}
+
+module Lens.Labels (
+    -- * Lenses
+    type (~>)(..),
+    LensLike,
+    LensLike',
+    (&),
+    (Category..),
+    Lens,
+    -- * HasLens
+    HasLens(..),
+    Proxy#(..),
+    proxy#,
+    -- * Setters
+    ASetter,
+    (.~),
+    (%~),
+    set,
+    over,
+    -- * Getters
+    Const(..),
+    Getting,
+    (^.),
+    view,
+    ) where
+
+import qualified Control.Category as Category
+import GHC.Prim (Proxy#, proxy#)
+#if __GLASGOW_HASKELL__ >= 800
+import GHC.OverloadedLabels (IsLabel(..))
+#endif
+import GHC.TypeLits (Symbol)
+
+import Data.Function ((&))
+#if __GLASGOW_HASKELL__ >= 800
+import Data.Functor.Const
+#endif
+import Data.Functor.Identity
+
+
+-- | A newtype for defining lenses.  Can be composed using
+-- `(Control.Category..)` (also exported from this module).
+newtype (~>) a b = LensFn {runLens :: a -> b}
+
+type LensFn = (~>)
```

Comment:
Is ~> prior art or is it irresistibly beautiful? :)
I'm just not sure if this is easy enough to parse compared to the more boring LensFn.